### PR TITLE
Fall back to old CA schema when retrieving keys and certs

### LINKED ISSUE #7576: Cannot log into cluster using Github with tsh.exe `v7.0.0-beta.1`
### Description

**What happened**: Logging into a Teleport cluster fails with tsh.exe `v7.0.0-beta.1`:

```
C:\Users\Gus>tsh-v7.0.0-beta.1.exe login --proxy=teleport.example.com --debug
←[37mDEBU←[0m [CLIENT]    Failed to stat file: CreateFile C:\Users\Gus\.tsh: The system cannot find the file specified.. client\api.go:679
←[37mDEBU←[0m [TSH]       Web proxy port was not set. Attempting to detect port number to use. tsh\tsh.go:1858
←[37mDEBU←[0m [TSH]       Resolving default proxy port (insecure: false) tsh\resolve_default_addr.go:121
←[37mDEBU←[0m [TSH]       Trying teleport.example.com:3080... tsh\resolve_default_addr.go:109
←[37mDEBU←[0m [TSH]       Trying teleport.example.com:443... tsh\resolve_default_addr.go:109
←[37mDEBU←[0m [TSH]       Address teleport.example.com:443 succeeded. Selected as canonical proxy address tsh\resolve_default_addr.go:195
←[37mDEBU←[0m [TSH]       Waiting for all in-flight racers to finish tsh\resolve_default_addr.go:144
←[37mDEBU←[0m [TSH]       Race request failed error:[Get "https://teleport.example.com:3080/webapi/ping": context canceled] tsh\resolve_default_addr.go:83
←[36mINFO←[0m [CLIENT]    No teleport login given. defaulting to VENUS\Gus client\api.go:1023
←[36mINFO←[0m [CLIENT]    no host login given. defaulting to VENUS\Gus client\api.go:1033
←[31mERRO←[0m [CLIENT]    [KEY AGENT] Unable to connect to SSH agent on socket: "". client\api.go:2740
←[37mDEBU←[0m [CLIENT]    not using loopback pool for remote proxy addr: teleport.example.com:443 client\api.go:2705
←[37mDEBU←[0m [CLIENT]    not using loopback pool for remote proxy addr: teleport.example.com:443 client\api.go:2705
←[37mDEBU←[0m [CLIENT]    HTTPS client init(proxyAddr=teleport.example.com:443, insecure=false) client\weblogin.go:212
←[36mINFO←[0m [CLIENT]    Waiting for response at: http://127.0.0.1:61400. client\redirect.go:137
If browser window does not open automatically, open it by clicking on the link:
 http://127.0.0.1:61400/5bb1a57b-46f7-411f-b6c2-fc08af5645b0
←[37mDEBU←[0m [CLIENT]    Got response from browser. client\weblogin.go:305
←[37mDEBU←[0m [KEYAGENT]  Adding CA key for teleport.example.com client\keyagent.go:289
←[37mDEBU←[0m [KEYSTORE]  Adding known host teleport.example.com with key: SHA256:1paZOuzvmf376q20ZZtfmedb0tS7yYGAGoetMh572aQ client\keystore.go:539
←[31mERRO←[0m [KEYSTORE]  open C:\Users\Gus\.tsh\keys\teleport.example.com\webvictim: The system cannot find the file specified. client\keystore.go:243
←[36mINFO←[0m [KEYAGENT]  Loading SSH key for user "webvictim" and cluster "teleport.example.com". client\keyagent.go:167
←[36mINFO←[0m [CLIENT]    Connecting proxy=teleport.example.com:3023 login="gus" client\api.go:2041
←[37mDEBU←[0m [KEYAGENT]  Validated host teleport.example.com:3023. client\keyagent.go:330
←[36mINFO←[0m [CLIENT]    Successful auth with proxy teleport.example.com:3023. client\api.go:2049
←[37mDEBU←[0m [KEYSTORE]  Returning Teleport TLS certificate "C:\\Users\\Gus\\.tsh\\keys\\teleport.example.com\\webvictim-x509.pem" valid until "2021-07-16 21:32:44 +0000 UTC". client\keystore.go:280
←[37mDEBU←[0m [CLIENT]    not using caching access point client\api.go:1132
←[37mDEBU←[0m [CLIENT]    Client  is connecting to auth server on cluster "teleport.example.com". client\client.go:792
←[37mDEBU←[0m [CLIENT]    Client  is connecting to auth server on cluster "teleport.example.com". client\client.go:792
←[37mDEBU←[0m [KEYAGENT]  Adding CA key for gus-tftestkube6 client\keyagent.go:289
←[37mDEBU←[0m [KEYAGENT]  Adding CA key for teleport.example.com client\keyagent.go:289
←[37mDEBU←[0m [KEYSTORE]  Returning Teleport TLS certificate "C:\\Users\\Gus\\.tsh\\keys\\teleport.example.com\\webvictim-x509.pem" valid until "2021-07-16 21:32:44 +0000 UTC". client\keystore.go:280
←[36mINFO←[0m [CLIENT]    Connecting proxy=teleport.example.com:3023 login="gus" client\api.go:2041
←[37mDEBU←[0m [KEYAGENT]  Validated host teleport.example.com:3023. client\keyagent.go:330
←[36mINFO←[0m [CLIENT]    Successful auth with proxy teleport.example.com:3023. client\api.go:2049
←[37mDEBU←[0m [CLIENT]    Found clusters: [{"name":"teleport.example.com","lastconnected":"2021-07-16T10:32:44.263879801-03:00","status":"online"}] client\client.go:109
←[37mDEBU←[0m [KEYSTORE]  Returning Teleport TLS certificate "C:\\Users\\Gus\\.tsh\\keys\\teleport.example.com\\webvictim-x509.pem" valid until "2021-07-16 21:32:44 +0000 UTC". client\keystore.go:280
←[37mDEBU←[0m [CLIENT]    Client  is connecting to auth server on cluster "teleport.example.com". client\client.go:792
←[37mDEBU←[0m [CLIENT]    Client  is connecting to auth server on cluster "teleport.example.com". client\client.go:792

ERROR REPORT:
Original Error: *trace.ConnectionProblemError Get &#34;https://teleport.cluster.local/v2/configuration/name&#34;: x509: certificate signed by unknown authority
Stack Trace:
        /go/src/github.com/gravitational/teleport/lib/httplib/httplib.go:133 github.com/gravitational/teleport/lib/httplib.ConvertResponse
        /go/src/github.com/gravitational/teleport/lib/auth/clt.go:307 github.com/gravitational/teleport/lib/auth.(*Client).Get
        /go/src/github.com/gravitational/teleport/lib/auth/clt.go:1544 github.com/gravitational/teleport/lib/auth.(*Client).GetClusterName
        /go/src/github.com/gravitational/teleport/tool/tsh/kube.go:260 main.fetchKubeClusters.func1
        /go/src/github.com/gravitational/teleport/lib/client/api.go:454 github.com/gravitational/teleport/lib/client.RetryWithRelogin
        /go/src/github.com/gravitational/teleport/tool/tsh/kube.go:248 main.fetchKubeClusters
        /go/src/github.com/gravitational/teleport/tool/tsh/kube.go:296 main.fetchKubeStatus
        /go/src/github.com/gravitational/teleport/tool/tsh/kube.go:353 main.updateKubeConfig
        /go/src/github.com/gravitational/teleport/tool/tsh/tsh.go:839 main.onLogin
        /go/src/github.com/gravitational/teleport/tool/tsh/tsh.go:583 main.Run
        /go/src/github.com/gravitational/teleport/tool/tsh/tsh.go:261 main.main
        /opt/go/src/runtime/proc.go:225 runtime.main
        /opt/go/src/runtime/asm_amd64.s:1371 runtime.goexit
User Message: Get &#34;https://teleport.cluster.local/v2/configuration/name&#34;: x509: certificate signed by unknown authority

C:\Users\Gus>
```

The browser indicated success (I'm using Github SSO on this cluster):

image

**What you expected to happen**: Logging in should work. The same command works with tsh.exe `v6.2.7`, so this appears to be a regression:

```
C:\Users\Gus>tsh-v6.2.7.exe version
Teleport v6.2.7 git:v6.2.7-0-g4f8a0007f go1.15.5

C:\Users\Gus>tsh-v6.2.7.exe login --proxy=teleport.example.com --debug
←[37mDEBU←[0m [CLIENT]    open C:\Users\Gus\.tsh\teleport.example.com.yaml: The system cannot find the file specified. client\api.go:710
←[37mDEBU←[0m [TSH]       Web proxy port was not set. Attempting to detect port number to use. tsh\tsh.go:1821
←[37mDEBU←[0m [TSH]       Resolving default proxy port (insecure: false) tsh\resolve_default_addr.go:121
←[37mDEBU←[0m [TSH]       Trying teleport.example.com:3080... tsh\resolve_default_addr.go:109
←[37mDEBU←[0m [TSH]       Trying teleport.example.com:443... tsh\resolve_default_addr.go:109
←[37mDEBU←[0m [TSH]       Address teleport.example.com:443 succeeded. Selected as canonical proxy address tsh\resolve_default_addr.go:195
←[37mDEBU←[0m [TSH]       Race request failed error:[Get "https://teleport.example.com:3080/webapi/ping": context canceled] tsh\resolve_default_addr.go:83
←[37mDEBU←[0m [TSH]       Waiting for all in-flight racers to finish tsh\resolve_default_addr.go:144
←[36mINFO←[0m [CLIENT]    No teleport login given. defaulting to VENUS\Gus client\api.go:1025
←[36mINFO←[0m [CLIENT]    no host login given. defaulting to VENUS\Gus client\api.go:1035
←[31mERRO←[0m [CLIENT]    [KEY AGENT] Unable to connect to SSH agent on socket: "". client\api.go:2741
←[37mDEBU←[0m [CLIENT]    not using loopback pool for remote proxy addr: teleport.example.com:443 client\api.go:2706
←[37mDEBU←[0m [CLIENT]    not using loopback pool for remote proxy addr: teleport.example.com:443 client\api.go:2706
←[37mDEBU←[0m [CLIENT]    HTTPS client init(proxyAddr=teleport.example.com:443, insecure=false) client\weblogin.go:211
←[36mINFO←[0m [CLIENT]    Waiting for response at: http://127.0.0.1:52809. client\redirect.go:137
If browser window does not open automatically, open it by clicking on the link:
 http://127.0.0.1:52809/e6797828-a104-4a7a-ad42-7a03a0e93aae
←[37mDEBU←[0m [CLIENT]    Got response from browser. client\weblogin.go:304
←[37mDEBU←[0m [KEYAGENT]  Adding CA key for teleport.example.com client\keyagent.go:289
←[37mDEBU←[0m [KEYSTORE]  Adding known host teleport.example.com with key: SHA256:1paZOuzvmf376q20ZZtfmedb0tS7yYGAGoetMh572aQ client\keystore.go:537
←[37mDEBU←[0m [KEYSTORE]  Returning Teleport TLS certificate "C:\\Users\\Gus\\.tsh\\keys\\teleport.example.com\\webvictim-x509.pem" valid until "2021-07-16 21:24:04 +0000 UTC". client\keystore.go:318
←[37mDEBU←[0m [KEYAGENT]  Deleting obsolete stored key with index {ProxyHost:teleport.example.com Username:webvictim ClusterName:teleport.example.com}. client\keyagent.go:437
←[36mINFO←[0m [KEYAGENT]  Loading SSH key for user "webvictim" and cluster "teleport.example.com". client\keyagent.go:167
←[36mINFO←[0m [CLIENT]    Connecting proxy=teleport.example.com:3023 login="gus" client\api.go:2043
←[37mDEBU←[0m [KEYAGENT]  Validated host teleport.example.com:3023. client\keyagent.go:330
←[36mINFO←[0m [CLIENT]    Successful auth with proxy teleport.example.com:3023. client\api.go:2051
←[37mDEBU←[0m [KEYSTORE]  Returning Teleport TLS certificate "C:\\Users\\Gus\\.tsh\\keys\\teleport.example.com\\webvictim-x509.pem" valid until "2021-07-16 21:32:02 +0000 UTC". client\keystore.go:318
←[37mDEBU←[0m [CLIENT]    not using caching access point client\api.go:1134
←[37mDEBU←[0m [CLIENT]    Client  is connecting to auth server on cluster "teleport.example.com". client\client.go:780
←[37mDEBU←[0m [CLIENT]    Client  is connecting to auth server on cluster "teleport.example.com". client\client.go:780
←[37mDEBU←[0m [KEYAGENT]  Adding CA key for gus-tftestkube6 client\keyagent.go:289
←[37mDEBU←[0m [CLIENT]    Client  is connecting to auth server on cluster "teleport.example.com". client\client.go:780
←[37mDEBU←[0m [KEYSTORE]  Adding known host gus-tftestkube6 with key: SHA256:CPUuUqYECQm4fjTASwbI1kIh86YeNvS2z62w6g4ihyM client\keystore.go:537
←[37mDEBU←[0m [KEYAGENT]  Adding CA key for teleport.example.com client\keyagent.go:289
←[37mDEBU←[0m [KEYSTORE]  Adding known host teleport.example.com with key: SHA256:1paZOuzvmf376q20ZZtfmedb0tS7yYGAGoetMh572aQ client\keystore.go:537
←[37mDEBU←[0m [KEYSTORE]  Returning Teleport TLS certificate "C:\\Users\\Gus\\.tsh\\keys\\teleport.example.com\\webvictim-x509.pem" valid until "2021-07-16 21:32:02 +0000 UTC". client\keystore.go:318
←[36mINFO←[0m [CLIENT]    Connecting proxy=teleport.example.com:3023 login="gus" client\api.go:2043
←[37mDEBU←[0m [KEYAGENT]  Validated host teleport.example.com:3023. client\keyagent.go:330
←[36mINFO←[0m [CLIENT]    Successful auth with proxy teleport.example.com:3023. client\api.go:2051
←[37mDEBU←[0m [CLIENT]    Found clusters: [{"name":"teleport.example.com","lastconnected":"2021-07-16T10:32:02.150567497-03:00","status":"online"}] client\client.go:109
←[37mDEBU←[0m [KEYSTORE]  Returning Teleport TLS certificate "C:\\Users\\Gus\\.tsh\\keys\\teleport.example.com\\webvictim-x509.pem" valid until "2021-07-16 21:32:02 +0000 UTC". client\keystore.go:318
←[37mDEBU←[0m [CLIENT]    Client  is connecting to auth server on cluster "teleport.example.com". client\client.go:780
←[36mINFO←[0m [CLIENT]    Connecting proxy=teleport.example.com:3023 login="gus" client\api.go:2043
←[37mDEBU←[0m [KEYAGENT]  Validated host teleport.example.com:3023. client\keyagent.go:330
←[36mINFO←[0m [CLIENT]    Successful auth with proxy teleport.example.com:3023. client\api.go:2051
←[37mDEBU←[0m [KEYSTORE]  Returning Teleport TLS certificate "C:\\Users\\Gus\\.tsh\\keys\\teleport.example.com\\webvictim-x509.pem" valid until "2021-07-16 21:32:02 +0000 UTC". client\keystore.go:318
←[37mDEBU←[0m [KEYSTORE]  Returning Teleport TLS certificate "C:\\Users\\Gus\\.tsh\\keys\\teleport.example.com\\webvictim-x509.pem" valid until "2021-07-16 21:32:02 +0000 UTC". client\keystore.go:318
←[37mDEBU←[0m [CLIENT]    Client  is connecting to auth server on cluster "teleport.example.com". client\client.go:780
←[37mDEBU←[0m [KEYSTORE]  Returning Teleport TLS certificate "C:\\Users\\Gus\\.tsh\\keys\\teleport.example.com\\webvictim-x509.pem" valid until "2021-07-16 21:32:02 +0000 UTC". client\keystore.go:318
←[37mDEBU←[0m [KEYSTORE]  Reading certificates from path "C:\\Users\\Gus\\.tsh\\keys\\teleport.example.com\\webvictim-ssh\\teleport.example.com-cert.pub". client\keystore.go:341
←[37mDEBU←[0m [KEYSTORE]  Returning Teleport TLS certificate "C:\\Users\\Gus\\.tsh\\keys\\teleport.example.com\\webvictim-x509.pem" valid until "2021-07-16 21:32:02 +0000 UTC". client\keystore.go:318
←[37mDEBU←[0m [KEYSTORE]  Reading certificates from path "C:\\Users\\Gus\\.tsh\\keys\\teleport.example.com\\webvictim-ssh\\teleport.example.com-cert.pub". client\keystore.go:341
> Profile URL:        https://teleport.example.com:443
  Logged in as:       webvictim
  Cluster:            teleport.example.com
  Roles:              admin, allow_root
  Traits:             kubernetes_groups: []
                      kubernetes_users: []
                      logins: [webvictim]
  Logins:             gus, webvictim, ec2-user, bots, root
  Kubernetes:         enabled
  Kubernetes cluster: "gus-gke-test-teleport-merged"
  Kubernetes users:   *
  Kubernetes groups:  system:masters
  Valid until:        2021-07-16 18:32:01 -0300 ADT [valid for 8h0m0s]
  Extensions:         permit-agent-forwarding, permit-port-forwarding, permit-pty


C:\Users\Gus>
```

I tried `tsh logout` and deleting `C:\Users\Gus\.tsh`, problem persists.

### Reproduction Steps
Attempt to log into a cluster with tsh.exe `v7.0.0-beta.1`. Observe failure.

### Server Details
- Teleport version (run `teleport version`): `Teleport Enterprise v6.2.3 git:v6.2.3-0-g3f2dd006c go1.15.5`
- Server OS (e.g. from `/etc/os-release`): `Ubuntu 20.04.2 LTS`
- Where are you running Teleport? (e.g. AWS, GCP, Dedicated Hardware): Home lab

### Client Details
- Tsh version (`tsh version`): `Teleport v7.0.0-beta.1 git:v7.0.0-beta.1-0-g29c390f1f go1.16.2`
- Computer OS (e.g. Linux, macOS, Windows): `Microsoft Windows Version 20H2 (OS Build 19042.1052)`

### PR DESCRIPTION
**Pull request by nklaassen**
fixes https://github.com/gravitational/teleport/issues/7576 introduced in https://github.com/gravitational/teleport/pull/7245

With this PR `GetActiveKeys`, `GetAdditionalTrustedKeys`, `GetTrustedSSHKeyPairs`, `GetTrustedTLSKeyPairs`, and `GetTrustedJWTKeyPairs` will all fall back to the "old" CA schema if the new `ActiveKeys` or `AdditionalTrustedKeys` fields are not set. This allows new version of `tsh` to work with old versions of `teleport` where the CA has not been migrated.

I will backport this to `branch/v7`

### CODE REVIEW

File: api/types/authority.go [hunk 1]
```diff
@@ -288,15 +288,64 @@ func (ca *CertAuthorityV2) SetSigningAlg(alg CertAuthoritySpecV2_SigningAlgType)
...
 func (ca *CertAuthorityV2) SetActiveKeys(ks CAKeySet) error {
 	if err := ks.CheckAndSetDefaults(); err != nil {
 		return trace.Wrap(err)
 	}
 	ca.Spec.ActiveKeys = ks
 	return nil
 }
-func (ca *CertAuthorityV2) GetAdditionalTrustedKeys() CAKeySet { return ca.Spec.AdditionalTrustedKeys }
+func (ca *CertAuthorityV2) GetAdditionalTrustedKeys() CAKeySet {
+	return ca.Spec.AdditionalTrustedKeys
```
Comment (timothyb89) [inline api/types/authority.go:324]: Looks like this return should be removed?

Comment (nklaassen) [reply api/types/authority.go]: right, woops. Should be fixed.

File: lib/services/authority_test.go [hunk 2]
```diff
@@ -76,11 +89,16 @@ func TestCertPoolFromCertAuthorities(t *testing.T) {
 		require.NoError(t, err)
 		require.Len(t, pool.Subjects(), 2)
 	})
+	t.Run("ca3 with 3 certs", func(t *testing.T) {
```
Comment (Joerger) [inline lib/services/authority_test.go:92]:
```suggestion
	t.Run("ca3 with 1 cert", func(t *testing.T) {
```

File: lib/services/authority_test.go [hunk 3]
```diff
@@ -66,6 +66,19 @@ func TestCertPoolFromCertAuthorities(t *testing.T) {
 	})
 	require.NoError(t, err)
 
+	// CA for cluster3 with old schema
+	key, cert, err = tlsca.GenerateSelfSignedCA(pkix.Name{CommonName: "cluster3"}, nil, time.Minute)
+	require.NoError(t, err)
+	ca3, err := types.NewCertAuthority(types.CertAuthoritySpecV2{
+		Type:        types.HostCA,
+		ClusterName: "cluster1",
```
Comment (Joerger) [inline lib/services/authority_test.go:74]:
```suggestion
		ClusterName: "cluster3",
```

File: api/types/authority.go [hunk 4]
```diff
@@ -288,15 +288,66 @@ func (ca *CertAuthorityV2) SetSigningAlg(alg CertAuthoritySpecV2_SigningAlgType)
...
+		}
+		keySet.SSH = []*SSHKeyPair{kp}
+	}
+	if len(ca.Spec.TLSKeyPairs) > 0 {
+		keySet.TLS = []*TLSKeyPair{ca.Spec.TLSKeyPairs[0].Clone()}
+	}
+	if len(ca.Spec.JWTKeyPairs) > 0 {
+		keySet.JWT = []*JWTKeyPair{ca.Spec.JWTKeyPairs[0].Clone()}
+	}
+	return keySet
```
Comment (Joerger) [inline api/types/authority.go:314]: Can you move this logic to a function like `func (ca *CertAuthorityV2) getOldKeySet(index uint) CaKeySet` and use that in `GetActiveKeys` and `GetAdditionalTrustedKeys`?

**Review by Joerger — APPROVED** [review]

**Review by xacrimon — APPROVED** [review]

File: api/types/authority.go (hunk 4 above)
Comment (nklaassen) [reply api/types/authority.go]: Good idea, done in https://github.com/gravitational/teleport/pull/7603/commits/8d83e6b5d57ab77a747a231e39e52c03e470c132

**Review by russjones — APPROVED** [review]
Bot.

**Pull request merged**